### PR TITLE
WIP: Add dynamic tracing system to JS library code

### PR DESCRIPTION
**Pull request by sbc100**
This system is designed to replace all the existing compile time DEBUG   
settings (SYSCALL_DEBUG, GL_DEBUG, FS_DEBUG, etc) with a system where       
the channels can be enabled/disabled at runtime.                            
                                                                            
The old settings till work and enable the respective channel at startup. 
                                                                            
Under node channels can be enabled by running with EM_DEBUG set in the   
environment.  This is a comma separated list of channel names.     

Fixes: #15057

### CODE REVIEW

**Comment by sbc100** [comment]
WTDY think about this @kripken, @juj. 

I've been wanting to be able to turn on an off trace channels without recompiling for while and this serves that purpose.

File: src/library_sockfs.js [hunk 1]
```diff
@@ -34,13 +34,13 @@ mergeInto(LibraryManager.library, {
...
-      Module['websocket']['on']('connection', function(fd) {out('Socket connection fd = ' + fd);});
-      Module['websocket']['on']('message', function(fd) {out('Socket message fd = ' + fd);});
-      Module['websocket']['on']('close', function(fd) {out('Socket close fd = ' + fd);});
+#if TRACING
+      Module['websocket']['on']('error', function(error) {trace(SOCKET, 'Socket error ' + error);});
+      Module['websocket']['on']('open', function(fd) {trace(SOCKET, 'Socket open fd = ' + fd);});
+      Module['websocket']['on']('listen', function(fd) {trace(SOCKET, 'Socket listen fd = ' + fd);});
+      Module['websocket']['on']('connection', function(fd) {trace(SOCKET, 'Socket connection fd = ' + fd);});
+      Module['websocket']['on']('message', function(fd) {trace(SOCKET, 'Socket message fd = ' + fd);});
+      Module['websocket']['on']('close', function(fd) {trace(SOCKET, 'Socket close fd = ' + fd);});
```
Comment (juj) [inline src/library_sockfs.js:43]: Missing quotes around `SOCKET` here

File: src/library_webgl.js [hunk 2]
```diff
@@ -574,7 +570,7 @@ var LibraryGL = {
     // Returns the context handle to the new context.
     createContext: function(canvas, webGLContextAttributes) {
 #if OFFSCREEN_FRAMEBUFFER
-      // In proxied operation mode, rAF()/setTimeout() functions do not delimit frame boundaries, so can't have WebGL implementation
+      // In proxied operation mode, rAF()/setTimetrace('GL', ) functions do not delimit frame boundaries, so can't have WebGL implementation
```
Comment (juj) [inline src/library_webgl.js:573]: search-replace typo here

**Comment by juj** [comment]
I do appreciate the idea, but I think this takes the meaning of these flags to a wrong direction. Compile time flags seem like a better design and more flexibility imo.

1) previously it was possible to enable/disable individual subsystem deep debugging one subsystem at a time. That matches what one typically wants to do: most often one has either a GL problem, or a pthreads problem, etc. so one knows "I want to enable verbose debugging for subsystem X". So I don't think there is a big need to do this kind of "enable verbose debugging on everything vs nothing." (and nothing prevented one to pass multiple flags in the build before either)
2) the meaning of the deep debugging flags has been exactly that, to add more involved debugging checks to the build. It is not just about tracing execution/logging. The meaning is mutated after this change.
3) the deep debugging flags can bring potentially runtime costly work to the build that can greatly slow down execution. That was efficiently mitigated by allowing one to enable only the ones that one needs. The new `TRACING` flag would enable deep debugging on everything at the same time, which would lose us the fine grained possibility to do e.g. deep validation/debugging on some pthreads behavior.
4) likewise, the deep debugging flags could bring in a lot of new added code to the build, but one could control it per subsystem. With `TRACING`, the build sizes can blow up unintentionally on tons of other subsystems that one doesn't care about.
5) there is a great advantage to new developers to have runtime boilerplate system be as low as possible. This new tracing system brings in a logging manager, which was not really needed before. It seems over-engineered for this purpose, and it means that developers have more core runtime code to read through (and more core concepts to learn to understand how to author or debug system runtime JS libraries - the previous system was more self-explanatory/bare JS).

I appreciate that a general tracing system seems like something better since it looks generalized from specific cases. But I think that is fallacious here. The existing set of compile-time flags `XXX_DEBUG` already had equally generalized meaning - sure it is from a compile-time set of "soft" conventions - but there is really no need to turn it into an enforced/engineered machinery for it to fit its purpose? It seems like an illusion of generalization to me.

Ultimately, I don't think this PR brings in any features that would be critical(?), but builds up an abstraction, with caveats mentioned above. The runtime toggling of enabled channels per build is the only distinct feature, but that does not seem like anyone would actually need it to debug (the scope is for deep debugging scenarios, not for logging channels scenarios). Not sure what prompted #15057, but I think we should not do this :(

**Pull request closed without merging**